sklearn/utils/testing.py

```diff
@@ -398,6 +407,179 @@ def assert_raise_message(exception, message, function, *args, **kwargs):
+    return False  # if the method does not exist.
+
+
+def assert_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if models are same"""
+    _assert_same_model_method('predict', X, estimator1, estimator2, msg)
+    _assert_same_model_method('transform', X, estimator1, estimator2, msg)
+    _assert_same_model_method('decision_function',
+                              X, estimator1, estimator2, msg)
+    _assert_same_model_method('predict_proba',
+                              X, estimator1, estimator2, msg)
+
+    assert_attributes_equal(estimator1, estimator2)
+
+
+def assert_not_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if models are different"""
+    t1 = _not_same_model_method('predict', X, estimator1, estimator2)
+    t2 = _not_same_model_method('transform', X, estimator1, estimator2)
+    t3 = _not_same_model_method('decision_function', X, estimator1, estimator2)
+    t4 = _not_same_model_method('predict_proba', X, estimator1, estimator2)
+
+    if msg is None:
+        msg = ("Models are equal. Both models returned the same results for"
+               " all the methods.")
+
+    any_method_differed = any((t1, t2, t3, t4))
+
+    if not any_method_differed:
+        # If all methods return similar results for both ests or if those
+        # methods are not present
+        assert_attributes_not_equal(estimator1, estimator2)
+
+
+def _compare_attributes(estimator1, estimator2, _invert_logic=False):
+    """Helper function to check if fitted model attributes are equal or not
+
+    Raises AssertionError if the attributes are equal ( or not equal, if
+    _invert_logic is True )
+    """
+    # A list of attributes which are known to be inconsistent.
+    skip_attributes = ('embedding_', 'weights_')
+    attr_dict = {}
+    est1_dict = estimator1.__dict__
+    est2_dict = estimator2.__dict__
+    # TODO Add more comparables
+    # TODO Support sparse matrices?
+    comparables = [list, np.ndarray, np.float64, float, int, 'str']
+
+    if not _invert_logic:
+        assert_equality = assert_equal
+        assert_array_equality = assert_array_equal
+    else:
+        assert_equality = assert_not_equal
+        assert_array_equality = assert_array_not_equal
+
+    # Collect all attributes from both the models
+    for e_dict in (est1_dict, est2_dict):
+        for key, value in e_dict.items():
+            if (key.endswith('_') and not key.endswith('estimator_') and
+                    not key.endswith('estimators_') and
+                    type(value) in comparables and
+                    key not in skip_attributes):
+
+                attr_dict[key] = type(value)
+
+    for attr, typ in attr_dict.items():
+        # Check equality of the attributes
+        try:
+            # If the attr types differ
+            if type(est1_dict[attr]) is not type(est2_dict[attr]):
+                raise KeyError("Both attributes are not of similar type")
```

Comment:
Actually the message inside `KeyError` was a bit misleading sorry. This was intended to be caught by the `try` / `except` block.

i.e if the attr are of different type : 
 - we need to make sure model is same -> we raise an error at this point itself.
 - we need to make sure model is different -> we assert True without raising anything.

`KeyError` since est1_dict[attr] at L542, can also raise `KeyError` when `attr` is not present.

Do you feel this is convoluted and should perhaps be refactored?